cleveref-usedon.dtx

```diff
@@ -579,33 +617,114 @@
 %	This |clist| contains all options that can be passed to |\cref| which are currently implemented.
 %    \begin{macrocode}
 \clist_new:N \g_@@_Options_clist
-\clist_set:Nn \g_@@_Options_clist {UsedOn, UsedBy, UsedByAndOn}
+\clist_set:Nn \g_@@_Options_clist { UsedOn, UsedBy, UsedByAndOn }
 %    \end{macrocode}
 %\end{variable}
 %
+%    \begin{macrocode}
+\cs_new:Nn \@@_UsedOnMessage:n { }
+\cs_new:Nn \@@_UsedByMessage:n { }
+\cs_new:Nn \@@_UsedByAndOnMessage:nn { }
+\NewDocumentCommand \UsedOnText { m } { \@@_UsedOnMessage:n { #1 } }
+\NewDocumentCommand \UsedByText { m } { \@@_UsedByMessage:n { #1 } }
+\NewDocumentCommand \UsedByAndOnText { m } { \@@_UsedByAndOnMessage:n { #1 } }
+%    \end{macrocode}
+%
+%\begin{macro}{\SetUsedOnText}
+%
+%    \begin{macrocode}
+\NewDocumentCommand \SetUsedOnText { m }
+  {
+    \cs_set:Nn \@@_UsedOnMessage:n { #1 }
+  }
+%    \end{macrocode}
+%\end{macro}
+%
+%\begin{macro}{\SetUsedByText}
+%
+%    \begin{macrocode}
+\NewDocumentCommand \SetUsedByText { m }
+  {
+    \cs_set:Nn \@@_UsedByMessage:n { #1 }
+  }
+%    \end{macrocode}
+%\end{macro}
+%
+%\begin{macro}{\SetUsedByAndOnText}
+%
+%    \begin{macrocode}
+\NewDocumentCommand \SetUsedByAndOnText { m }
+  {
+    \cs_set:Nn \@@_UsedByAndOnMessage:nn { #1 }
+  }
+%    \end{macrocode}
+%\end{macro}
+%
+%    \begin{macrocode}
+\RequirePackage{iflang}
+
+\SetUsedOnText
+  {
+    \IfLanguageName{english}
+        { (Used~on~#1.) } {}
+    \IfLanguageName{french}
+        { (Apparaît~en~#1.) } {}
+    \IfLanguageName{ngerman}
+        { (Wird~auf~#1.) } {}
+    \IfLanguageName{spanish}
+        { (Aparece~en~#1.) } {}
+  }
+\SetUsedByText
+  {
+    \IfLanguageName{english}
+        { (Used~by~#1.) } {}
+    \IfLanguageName{french}
+        { (Apparaît~dans~#1.) } {}
+    \IfLanguageName{ngerman}
+        { (Wird~in~#1.) } {}
+    \IfLanguageName{spanish}
+        { (Aparece~en~#1.) } {}
+  }
+\SetUsedByAndOnText
+  {
+    \IfLanguageName{english}
+        { (Used~by~#1~on~#2.) } {}
+    \IfLanguageName{french}
+        { (Apparaît~dans~#1~en~#2.) } {}
+    \IfLanguageName{ngerman}
+        { (Wird~in~#1~auf~#2.) } {}
+    \IfLanguageName{spanish}
+        { (Aparece~en~#1~en~#2.) } {}
+  }
+
+%    \end{macrocode}
+%
 %\begin{macro}{\UsedOnMessage}
 %	The following are the standard texts that get printed in the first line
 %	of the labelled environment which later gets referenced with
 %   |[UsedOn]|, |[UsedBy]| or |[UsedByAndOn]|.
 %
 %    \begin{macrocode}
-\NewDocumentCommand{\UsedOnMessage}{m}{
-    \emph{(Used~on~#1.)} \newline
-}
+\NewDocumentCommand \UsedOnMessage { m }
```

Comment:
I shall modify the commands' names accordingly.

There are already `\Set...` version for all these commands in the code of #12 , as discussed in #10 (though some of them have to wait until the next pull request to appear). With the new command names, it would look like:
```
\SetUsedByMessageText
  {%
    Referenced by: #1.%
  }
\SetUsedByMessageStyle
  {%
    \todo[%
      inline,
      size=\scriptsize,
      color=blue!40,
      backgroundcolor=white,
      textcolor=darkgray,
      bordercolor=darkgray,
      prepend]{%
      \textup{\UsedByMessageText{#1}}%
    }%
  }
```
(Ah, so by now `\UsedByMessageText` is not really an internal command, I see... I shall also define a `\UsedByMessageContent` for it.)

Do you think perhaps, instead of `\UsedByMessageText{#1}`, it would be preferable for users to simply write `#1`, and let it be the task of the `\Set...` command to fill in the corresponding text command? (To be honest I haven't yet come up with an idea how to write the code of such a `\Set...` command.)

Reply:
Another thing.

In fact, the current `\UsedOnMessage` (to be renamed as `\UsedOnMessageStyle`) is for text mode. There are (or will be) also `\UsedOnMessageInsideMath` and `\UsedOnMessageOutsideMath`. Do you have suggestions for the name of them?

Reply:
Actually the expl3-level and user-level command should be separately defined — a practice I didn't follow quite well. For example, for the user level `\UsedOnMessage`, there should be an expl3-level `\@@_UsedOnMessage:n` first.

After this pull request, I shall proceed to include new features in the code in #12 in the next pull request, and after that, with all the code we already have at hand being integrated, I shall separately deal with this issue.